### ISSUE TITLE
EthereumStratum/1.0.0 mining.set_difficulty sets higher diff than real

### ISSUE THREAD

**Issue by AndreaLanfranchi** [ethereum-mining/ethminer#1681]
When EthereumStratum/1.0.0 is implemented by work provider (pool) the mining.set_difficulty pushes a difficulty index which is used by miner to set next job target.
Ethminer however sets the target to a higher difficulty thus reducing probability to find the right nonce.

**To Reproduce**

1. Connect miner to daggerhashimoto on NiceHash
2. Wait for a mining.set_difficulty notification 
3. With decimal compute target
4. Divide found target by base target at diff 1 `00000000ffff0000000000000000000000000000000000000000000000000000` and make the reciprocal. 
5. Value set in step 2 does not match with value obtained in step 4

**Expected behavior**
Values should match

**Sample**
NiceHash pushes diff `0.9313083637607632`
Ethminer computes target of `0000000112e0be82000000000000000000000000000000000000000000000000`
Now calculate
```
1 / (
    0000000112e0be82000000000000000000000000000000000000000000000000 /
    00000000ffff0000000000000000000000000000000000000000000000000000
    )
```
You get `0.999864959938281` which is a higher difficulty.

The correct target for diff `0.9313083637607632` should be
`0000000112e0be826d68384da44da7a8a5fbfd91e69cd791a1d379e51e52507f`

**Comment by StefanOberhumer** [comment]
@AndreaLanfranchi 
You're right!! (even never seen NiceHash pushes a difficulty of 0.9313083637607632)

Have tried with following playground 
```
#include <stdio.h>
#include <string.h>
#include <stdint.h>


static void diffToTarget(uint32_t* target, double diff)
{
	// target is in real life h256 type and is passed as 8 * uint32 array to this function
    uint32_t target2[8];
    uint64_t m;
    int k;


    for (k = 6; k > 0 && diff > 1.0; k--)
        diff /= 4294967296.0; // 0x100000000

    m = (uint64_t)(4294901760.0 / diff); // 0xffff0000 / diff
    if (m == 0 && k == 6)
        memset(target2, 0xff, 32);
    else
    {
        memset(target2, 0, 32);
        target2[k] = (uint32_t)m;
        target2[k + 1] = (uint32_t)(m >> 32);
    }

#if 1
	memcpy(target, target2, 8*4); // Test environment - better display
#else
	// reverse the bytes as target is type h256 (stored BE)
    for (int i = 0; i < 32; i++)
        ((uint8_t*)target)[31 - i] = ((uint8_t*)target2)[i];
#endif
}


void test(double diff)
{
	uint32_t t[8];

	diffToTarget(&t[0], diff);

	printf("0x%08x%08x%08x%08x%08x%08x%08x%08x %f\n", t[7],t[6],t[5],t[4],t[3],t[2],t[1],t[0], diff);
}

void main(void)
{
	test(1);
	test(0.5);
	test(0.9313083637607632);
}
```

It seems as we loose the part after the decimal point in the line ```m = (uint64_t)(4294901760.0 / diff);```

**Comment by StefanOberhumer** [comment]
Making it better but not at estimated target:
```
static void diffToTarget(uint32_t* target, double diff)
{
	// target is in real life h256 type and is passed as 8 * uint32 array to this function
    uint32_t target2[8];
    uint64_t m;
    double x;
    int k;

    for (k = 6; k > 0 && diff > 1.0; k--)
        diff /= 4294967296.0; // 0x100000000

	x = 4294901760.0 / diff; // 0xffff0000 / diff
    m = (uint64_t)(x);
    if (m == 0 && k == 6)
        memset(target2, 0xff, 32);
    else
    {
        memset(target2, 0, 32);
        target2[k] = (uint32_t)m;
        target2[k+1] = (uint32_t)(m >> 32);

		// care about after decimal point numbers
        x -= m;
        while (--k >= 0 && x != 0.0)
        {
			x *= 4294967296.0;
			printf("%f ",x);
			target2[k] = (uint32_t) x;
			x -= (uint32_t) x;
		}
    }

#if 1
	memcpy(target, target2, 8*4); // Test environment - better display
#else
	// reverse the bytes as h256 is stored BE
    for (int i = 0; i < 32; i++)
        ((uint8_t*)target)[31 - i] = ((uint8_t*)target2)[i];
#endif
}
```

**Comment by AndreaLanfranchi** [comment]
We're not quite there yet.
Your last mod produces :

`0x0000000112e0be826d6950000000000000000000000000000000000000000000`
which is better than
`0x0000000112e0be82000000000000000000000000000000000000000000000000`
but yet far from 
`0x0000000112e0be826d68384da44da7a8a5fbfd91e69cd791a1d379e51e52507f`

**Comment by MariusVanDerWijden** [comment]
I manage to get 
```
0x0000000112e0be826d6954680000000000000000000000000000000000000000
```
by using long double instead of doubles

**Comment by AndreaLanfranchi** [comment]
You mean long double both for x and diff ?

**Comment by AndreaLanfranchi** [comment]
Damn ... when shifting down bytes the multiplier must be 4294967295.0 (ffffffff) and not 4294967296.0 (100000000). Uint32 max value is (2^32-1)

**Comment by AndreaLanfranchi** [comment]
This gives a result of
`0x0000000112e0be826d694fff9296afff6d694fff9296afff6d694fff9296afff`
which is quite near to
`0x0000000112e0be826d68384da44da7a8a5fbfd91e69cd791a1d379e51e52507f`

**Comment by MariusVanDerWijden** [comment]
It still only shares the first 85 bits...

```
void diffFromResult(uint32_t* target)
{
    uint32_t t[8];
    t[7] = target[7] - 0x00000001;
    t[6] = target[6] - 0x12e0be82;
    t[5] = target[5] - 0x6d68384d;
    t[4] = target[4] - 0xa44da7a8;
    t[3] = target[3] - 0xa5fbfd91;
    t[2] = target[2] - 0xe69cd791;
    t[1] = target[1] - 0xa1d379e5;
    t[0] = target[0] - 0x1e52507f;
    printf("Diff: 0x%08x%08x%08x%08x%08x%08x%08x%08x\n", t[7],t[6],t[5],t[4],t[3],t[2],t[1],t[0]);
    clz(&t);
}



void clz(uint32_t* target)
{
    static const char debruijn32[32] = 
    {
        0, 31, 9, 30, 3, 8, 13, 29, 2, 5, 7, 21, 12, 24, 28, 19,
        1, 10, 4, 14, 6, 22, 25, 20, 11, 15, 23, 26, 16, 27, 17, 18
    };
    uint32_t lz = 0;
    for(int i = 0; i < 8; i++)
    {
        uint32_t x = target[i];
        if(!x)
        {
            lz += 32;
            continue;
        } 
        x |= x>>1;
        x |= x>>2;
        x |= x>>4;
        x |= x>>8;
        x |= x>>16;
        x++;
        lz += debruijn32[x*0x076be629>>27];
    }
    printf("Clz: %i\n",lz);
}
```

**Comment by AndreaLanfranchi** [comment]
So a long story.
I have implemented two tries using .Net's BigInteger structure.
First try using C# :
```
using System;
using System.Globalization;
using System.Numerics;

namespace ScratchPadCS
{
    public class Math
    {
        public string DiffToTarget(double diff)
        {

            // This is the target at difficulty 1
            BigInteger BaseTarget = BigInteger.Parse("00000000ffff0000000000000000000000000000000000000000000000000000", System.Globalization.NumberStyles.AllowHexSpecifier, null);

            // Make the reciprocal of diff
            diff = 1.0 / diff;

            // This is the new computed target
            // at first it's the multiplication of BaseTarget by integer part of diff
            BigInteger NewTarget;
            BigInteger DecimalDiff;
            BigInteger DecimalTarget;

            NewTarget = BigInteger.Multiply(BaseTarget, new BigInteger(diff));

            // Transform diff to string to capture decimal places
            string StringDiff = diff.ToString(System.Globalization.CultureInfo.InvariantCulture);
            int DecimalOffset = StringDiff.IndexOf(".");
            if (DecimalOffset > -1)
            {
                int Precision = (StringDiff.Length - 1) - DecimalOffset;
                DecimalDiff = BigInteger.Parse(StringDiff.Substring(DecimalOffset + 1));
                DecimalTarget = BigInteger.Multiply(BaseTarget, DecimalDiff);

                // For reasons beyond my comprehension this statement does not work
                // in C# while works perfectly in VB.Net
                // DecimalTarget = BigInteger.Divide(DecimalTarget, (10 ^ Precision));

                // Workaround using truncated string
                string s = DecimalTarget.ToString();
                s = s.Substring(0, s.Length - Precision);

                DecimalTarget = BigInteger.Parse(s);
                NewTarget += DecimalTarget;
            }

            return string.Format("{0:x64}", NewTarget);

        }

        public double TargetToDiff(string targetHex)
        {
            // Adjust to 64 chars
            targetHex = String.Format("{0}{1}", new String('0', 64), targetHex);
            targetHex = targetHex.Substring(targetHex.Length - 64);

            int Precision = 0;
            BigInteger Dividend;
            BigInteger Divisor;
            BigInteger Remainder;
            string Quotient;

            Precision = 0;
            Divisor = BigInteger.Parse("00000000ffff0000000000000000000000000000000000000000000000000000", System.Globalization.NumberStyles.AllowHexSpecifier, null);
            Dividend = BigInteger.Parse(targetHex, System.Globalization.NumberStyles.AllowHexSpecifier, null);
            Quotient = BigInteger.DivRem(Dividend, Divisor, out Remainder).ToString();
            if(Remainder != BigInteger.Zero)
            {
                Quotient += ".";
                while (Precision < 17 && Remainder != BigInteger.Zero)
                {
                    Precision++;
                    Dividend = BigInteger.Multiply(Remainder, new BigInteger(10 ^ Precision));
                    Quotient += BigInteger.DivRem(Dividend, Divisor, out Remainder).ToString();
                }
            }
            return 1 / Double.Parse(Quotient);
        }

    }


    class Program
    {

        static void Main(string[] args)
        {

            CultureInfo ci = new CultureInfo("en-US");
            CultureInfo.DefaultThreadCurrentCulture = ci;
            CultureInfo.DefaultThreadCurrentUICulture = ci;

            Math m = new Math();

            Console.WriteLine("Difficulty : 0.9313083637607632");
            Console.WriteLine("To Target  : {0}", m.DiffToTarget(0.9313083637607632));
            Console.WriteLine("Reverse    : {0}", m.TargetToDiff(m.DiffToTarget(0.9313083637607632)));
            Console.ReadLine();

        }

    }
}
```

Which eventually reports 
```
Difficulty : 0.9313083637607632
To Target  : 0000000112e0be826d68384da44da7a8a5fbfd91e69cd791a1d379e51e525084
Reverse    : 0.939397079229044
```

As you'll notice from an above comment the BigInteger.Divide method does not work in C# for very large numbers. So I decided to give VB.Net a try.

```
Imports System
Imports System.Globalization
Imports System.Numerics
Imports System.Threading

Module Program
    Sub Main(args As String())

        Dim ci As CultureInfo = New CultureInfo("en-US")
        CultureInfo.DefaultThreadCurrentCulture = ci
        CultureInfo.DefaultThreadCurrentUICulture = ci

        Console.WriteLine("Difficulty : 0.9313083637607632")
        Console.WriteLine("To Target  : {0}", DiffToTarget(0.9313083637607632))
        Console.WriteLine("Reverse    : {0}", TargetToDiff(DiffToTarget(0.9313083637607632)))

        Console.ReadKey()
    End Sub

    Public Function TargetToDiff(targetHex As String) As Double

        ' Adjust to 64 chars
        targetHex = String.Format("{0}{1}", New String("0", 64), targetHex)
        targetHex = targetHex.Substring(targetHex.Length - 64)

        Dim Precision As Integer = 0
        Dim Dividend As BigInteger = BigInteger.Zero
        Dim Divisor As BigInteger = BigInteger.Zero
        Dim Remainder As BigInteger = BigInteger.Zero
        Dim Quotient As String = String.Empty

        Precision = 0
        Divisor = BigInteger.Parse("00000000ffff0000000000000000000000000000000000000000000000000000", Globalization.NumberStyles.AllowHexSpecifier, Nothing)
        Dividend = BigInteger.Parse(targetHex, Globalization.NumberStyles.AllowHexSpecifier, Nothing)
        Quotient = BigInteger.DivRem(Dividend, Divisor, Remainder).ToString
        If Remainder <> BigInteger.Zero Then
            Quotient += "."
            While Precision < 17 AndAlso Remainder <> BigInteger.Zero
                Precision += 1
                Dividend = BigInteger.Multiply(Remainder, New BigInteger(10 ^ Precision))
                Quotient += BigInteger.DivRem(Dividend, Divisor, Remainder).ToString
            End While
        End If
        Return 1 / Double.Parse(Quotient)

    End Function

    Public Function DiffToTarget(diff As Double) As String

        diff = 1 / diff
        Dim base As BigInteger = BigInteger.Parse("00000000ffff0000000000000000000000000000000000000000000000000000", Globalization.NumberStyles.AllowHexSpecifier, Nothing)
        Dim product As BigInteger = BigInteger.Zero
        product = BigInteger.Multiply(base, New BigInteger(diff))

        Dim sdiff As String = diff.ToString
        Dim offset As Integer = sdiff.IndexOf(".")
        If offset > -1 Then
            Dim precision As Integer = (sdiff.Length - 1) - offset
            Dim newDiff As BigInteger = BigInteger.Parse(sdiff.Substring(offset + 1))
            newDiff = BigInteger.Multiply(base, newDiff)
            newDiff = BigInteger.Divide(newDiff, New BigInteger(10 ^ precision))
            product = BigInteger.Add(product, newDiff)
        End If
        Return String.Format("{0:x64}", product)

    End Function

End Module

```

And the result is 
```
Difficulty : 0.9313083637607632
To Target  : 0000000112e0be826d68384da44da7a8a5fbfd91e69cd791a1d379e51e525084
Reverse    : 0.931308363760767
```
__with a little glitch at the 15th decimal place__

Anyway in both cases the correct target for a diff `0.9313083637607632` is always
```
VB Target  : 0000000112e0be826d68384da44da7a8a5fbfd91e69cd791a1d379e51e525084
C# Target  : 0000000112e0be826d68384da44da7a8a5fbfd91e69cd791a1d379e51e525084
```

My originally posted target was                    `0000000112e0be826d68384da44da7a8a5fbfd91e69cd791a1d379e51e52507f` was 5 units below due to a problem in rounding decimals.

How we get those values in ethminer ?

**Comment by StefanOberhumer** [comment]
@AndreaLanfranchi  - Thanks for your C#/VB samples ... working on on a solutions - currently trying to use 

```
sprintf(buff, "%0'.64llf", x);
fractional = strtold(buff+2, NULL);
```

as its based on compiler implementation how "double" and "long double" are implemented.
And: AFAIK there is no BigNum in C++

Btw: It is a funny and interesting problem!

**Comment by AndreaLanfranchi** [comment]
Apparently someone took his time to build an implementation for BigInteger. Maybe it fits. Not tested yet.
https://github.com/panks/BigInteger

**Comment by StefanOberhumer** [comment]
@AndreaLanfranchi 

> Apparently someone took his time to build an implementation for BigInteger. Maybe it fits. Not tested yet. https://github.com/panks/BigInteger

Note sure if it fits as it seems dividend in this implementation has to fit into 'long long'

> // For reasons beyond my comprehension this statement does not work
>// in C# while works perfectly in VB.Net
>// DecimalTarget = BigInteger.Divide(DecimalTarget, (10 ^ Precision));

Could it be that the '10' makes the problem? (The term will be interpreted as 'int' and then casted to BigInt?)

**Comment by AndreaLanfranchi** [comment]
> Could it be that the '10' makes the problem? (The term will be interpreted as 'int' and then casted to BigInt?

Documentation states all members of the operation are implicitly casted to BigInteger.
Nevertheless I've done a bunch of tests with native BigIntegers, with other numbers ... division in C# does not work as expected.

**Comment by MariusVanDerWijden** [comment]
@StefanOberhumer 
I think GMP (https://gmplib.org/) is teh defacto standard library for big numbers in C/C++

**Comment by AndreaLanfranchi** [comment]
@MariusVanDerWijden @StefanOberhumer 
BigIntegers are implemented in boost too. Either as cpp_int or gmp_int.

I have not seen Marius implementation but probably he is right as only now I discovered the problems in rounding ToString.

I implemented it using cpp_int and finally got matching results. Here is the snippet.

```
static void diffToTarget3(double diff) {

    using namespace boost::multiprecision;
    using BigInteger = boost::multiprecision::cpp_int;

    static BigInteger base("0x00000000ffff0000000000000000000000000000000000000000000000000000");

    diff = 1 / diff;

    BigInteger idiff(diff);
    BigInteger product;
    product = base * idiff;

    std::string sdiff = boost::lexical_cast<std::string>(diff);
    size_t ldiff = sdiff.length();
    size_t offset = sdiff.find(".");

    if (offset != std::string::npos) 
    {
        // Number of decimal places
        size_t precision = (ldiff - 1) - offset;

        // Effective sequence of decimal places
        string decimals = sdiff.substr(offset + 1);

        // Uncomment the following lines if you wish to
        // compare the results with C#/VB samples.
        // In .Net, in fact, the reciprocal of
        // 0.9313083637607632 when turned into string
        // gets truncated to "1.07375820825" instead of
        // "1.0737582082500039"
        //decimals = "07375820825";
        //precision = 11;

        // Strip leading zeroes. If a string begins with
        // 0 or 0x boost parser considers it hex
        decimals = decimals.erase(0, decimals.find_first_not_of('0'));
        
        // Build up the divisor as string - just in case
        // parser does some implicit conversion with 10^precision
        string decimalDivisor = "1";
        decimalDivisor.resize(precision + 1, '0');

        // This is the multiplier for the decimal part        
        BigInteger multiplier(decimals);

        // This is the divisor for the decimal part
        BigInteger divisor(decimalDivisor);

        BigInteger decimalproduct;
        decimalproduct = base * multiplier;
        decimalproduct /= divisor;

        // Add the computed decimal part
        // to product
        product += decimalproduct;

    }

    // Normalize to 64 chars hex
    stringstream ss;
    ss << setw(64) << setfill('0') << std::hex << product;

    string target = ss.str();
    boost::algorithm::to_lower(target);
    std::cout << target << std::endl;

}
```

Which produces 
```
Difficulty: 0.9313083637607632
To target : 0000000112e0be826d695152d88fd99f116457c399d333a21486c46448c0d24c
```

If in the above VB.Net example you simply replace
```
Dim sdiff As String = diff.ToString
```
with
```
Dim sdiff As String = diff.ToString("R")
```
which forces to carry to string the maximum precision available you get
```
Difficulty : 0.9313083637607632
To Target  : 0000000112e0be826d695152d88fd99f116457c399d333a21486c46448c0d24c
Reverse    : 0.931308363760763
```

Full match.

**Comment by AndreaLanfranchi** [comment]
For instance : the method in use to use a bytewise division even on correction of the proper switcher `(2^32-1)` gives a different result `0x0000000112e0be826d694fff9296afff6d694fff9296afff6d694fff9296afff`
which is a higher target than it should be.

**Comment by AndreaLanfranchi** [comment]
Resolved by #1704
